src/twister2/device/native_simulator_adapter.py

```diff
@@ -23,78 +25,86 @@ def __init__(self, twister_config, hardware_map: HardwareMap | None = None, **kw
-                self._process = None
+    async def _run_command(self, command: list[str], timeout: float = 60):
+        assert isinstance(command, (list, tuple, set))  # to avoid stupid and  difficult to debug mistakes
+        self._process = await asyncio.create_subprocess_exec(
+            *command,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.STDOUT,
+            env=self.env
+        )
+        logger.debug('Started subprocess with PID %s', self._process.pid)
+        end_time = time.time() + timeout
+        while not self._stop_job:
+            try:
+                line = await asyncio.wait_for(self._process.stdout.readline(), 0.1)
+            except asyncio.TimeoutError:
+                pass
+            else:
+                if line:
+                    self.queue.put(line.decode('utf-8').strip())
+                    logger.debug(line.decode('utf-8').strip())
+            if time.time() > end_time:
+                self._process_ended_with_timeout = True
+                logger.debug(f'Finished process with PID {self._process.pid} with timeout')
+                break
+
+        return await self._process.wait()
 
-    def flash(self, build_dir: str | Path, timeout: float = 60.0) -> None:
-        """Run simulation."""
-        command: str = self.get_command(build_dir)
-        logger.info('Flashing device')
-        logger.info('Flashing command: %s', command)
+    def run(self) -> None:
+        command: list[str] = self._get_command(self.build_dir)
+        logger.info('Running command: %s', command)
         try:
-            self._process = subprocess.Popen(
-                command,
-                stdout=subprocess.PIPE,
-                stderr=subprocess.PIPE,
-                env=self.env,
+            self._loop = asyncio.new_event_loop()
+            return_code = self._loop.run_until_complete(
+                self._run_command(command, timeout=self.timeout)
             )
-
-            t1 = self._collect_process_output(self._process)
-            t2 = self._wait_and_terminate_process(self._process, timeout=timeout)
-            logger.info('Started process %s', self._process.pid)
-            t1.start()
-            t2.start()
-            t1.join()
-            t2.join()
-        except subprocess.CalledProcessError as e:
-            logger.exception('Flashing failed due to error %s', e)
-            raise
+        except subprocess.SubprocessError as e:
+            logger.error('Flashing failed due to subprocess error %s', e)
+            self._exc = TwisterFlashException(e.args)
+        except FileNotFoundError as e:
+            logger.error(f'Flashing failed due to file not found: {e.filename}')
+            self._exc = TwisterFlashException(f'File not found: {e.filename}')
+        except Exception as e:
+            logger.error('Flashing failed: %s', e)
+            self._exc = TwisterFlashException(e.args)
         else:
```

Comment:
There should be also verification of `self._process_ended_with_timeout` variable. Without this, now even if timeout occurs  test still pass without timeout failure. It can be considered if this timeout should be raised here or maybe in log parser:
https://github.com/zephyrproject-rtos/twister/blob/main/src/twister2/log_parser/ztest_log_parser.py#L41-L43

Reply:
We have to consider two scenarios:
1. We start simulator (subprocess) which prints some messages to stdout and ends. In test we check these messages to determine a status of this test. In this scenario we don't have to kill the process unless it got stuck for some some reason.  
2. We start simulator which prints some messages in a loop or just stops and waits. In this case we need to kill the process after test is ended. In test we are waiting e.g. for 10 seconds for expected messages and if those messages did not appear we consider the test status as fail. Now we need to run clean up and kill process which is still running. Should we raise an exception in tear down due to time out and set status for this test as error? Or just silently kill the process? (also I am not sure if we have to check in while loop if time out appeared if we do the same in test in log parser.

Reply:
Hmmmmm, in my opinion I think that it will be enough with checking timeout in parser and rise error only there. At this moment I don't see any reason to verify it additionally in `_run_command` method. At the same time we don't provide any mechanism to rise error/failure in log parser and this is a problem. I will try to write some proposition to cover this "gap" soon.